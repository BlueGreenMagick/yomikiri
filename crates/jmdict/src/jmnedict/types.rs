use serde::{Deserialize, Serialize};

use crate::jmdict::JMSenseMisc;
use crate::utils::jm_entity_enum;

#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneDict {
    pub entries: Vec<JMneEntry>,
    pub creation_date: Option<String>,
}

/// `<entry>`
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneEntry {
    pub id: u32,
    /// 0+ k_ele
    pub kanjis: Vec<JMneKanji>,
    /// 1+ r_ele
    pub readings: Vec<JMneReading>,
    /// 1+ trans
    pub translations: Vec<JMneTranslation>,
}

/// `<k_ele>`
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneKanji {
    pub kanji: String,
    // Defined, but not used.
    // pub info: Vec<String>,
    pub priority: Vec<String>,
}

/// `<r_ele>`
#[derive(Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct JMneReading {
    pub reading: String,
    pub to_kanji: Vec<String>,
    // Defined, but not used.
    // pub info: Vec<String>,
    pub priority: Vec<String>,
}

/// `<trans>`
///
/// At the date of development,
/// while specs allows translations in languages other than English,
/// no such translations actually exists in xml.
/// If non-English translations are added in the future,
/// they will be ignored and not be added to `translations` field.
#[derive(Debug, PartialEq, Eq, Default, Serialize, Deserialize)]
pub struct JMneTranslation {
    pub name_type: Vec<JMneNameType>,
    pub xref: Vec<String>,
    /// English translations are included
    pub translations: Vec<String>,
}

jm_entity_enum!(
    JMneNameType;
    #[doc="character"]
    b"char" => Character,
    #[doc="company name"]
    b"company" => Company,
    #[doc="creature"]
    b"creat" => Creature,
    #[doc="deity"]
    b"dei" => Deity,
    #[doc="document"]
    b"doc" => Document,
    #[doc="event"]
    b"ev" => Event,
    #[doc="female given name or forename"]
    #[serde(rename="forename (female)")]
    b"fem" => Female,
    #[doc="fiction"]
    b"fict" => Fiction,
    #[doc="given name or forename, gender not specified"]
    b"given" => Forename,
    #[doc="group"]
    b"group" => Group,
    #[doc="legend"]
    b"leg" => Legend,
    #[doc="male given name or forename"]
    #[serde(rename="forename (male)")]
    b"masc" => Male,
    #[doc="mythology"]
    b"myth" => Mythology,
    #[doc="object"]
    b"obj" => Object,
    #[doc="organization name"]
    b"organization" => Organization,
    #[doc="other"]
    b"oth" => Other,
    #[doc="full name of a particular person"]
    b"person" => Person,
    #[doc="place name"]
    b"place" => Place,
    #[doc="product name"]
    b"product" => Product,
    #[doc="religion"]
    b"relig" => Religion,
    #[doc="service"]
    b"serv" => Service,
    #[doc="ship name"]
    b"ship" => Ship,
    #[doc="railway station"]
    b"station" => Railway,
    #[doc="family or surname"]
    b"surname" => Surname,
    #[doc="unclassified name"]
    b"unclass" => Unclassified,
    #[doc="work of art, literature, music, etc. name"]
    b"work" => Artwork,
);

impl JMneReading {
    /// Returns true if reading is applicable to kanji
    ///
    /// `kanji` must be in same entry.
    pub fn is_for_kanji(&self, kanji: &String) -> bool {
        self.to_kanji.is_empty() || self.to_kanji.contains(kanji)
    }
}

impl From<JMneNameType> for JMSenseMisc {
    fn from(value: JMneNameType) -> Self {
        match value {
            JMneNameType::Artwork => JMSenseMisc::Artwork,
            JMneNameType::Character => JMSenseMisc::Character,
            JMneNameType::Company => JMSenseMisc::Company,
            JMneNameType::Creature => JMSenseMisc::Creature,
            JMneNameType::Deity => JMSenseMisc::Deity,
            JMneNameType::Document => JMSenseMisc::Document,
            JMneNameType::Event => JMSenseMisc::Event,
            JMneNameType::Female => JMSenseMisc::Female,
            JMneNameType::Fiction => JMSenseMisc::Fiction,
            JMneNameType::Forename => JMSenseMisc::Forename,
            JMneNameType::Group => JMSenseMisc::Group,
            JMneNameType::Legend => JMSenseMisc::Legend,
            JMneNameType::Male => JMSenseMisc::Male,
            JMneNameType::Mythology => JMSenseMisc::Mythology,
            JMneNameType::Object => JMSenseMisc::Object,
            JMneNameType::Organization => JMSenseMisc::Organization,
            JMneNameType::Other => JMSenseMisc::Other,
            JMneNameType::Person => JMSenseMisc::Person,
            JMneNameType::Place => JMSenseMisc::Place,
            JMneNameType::Product => JMSenseMisc::Product,
            JMneNameType::Railway => JMSenseMisc::Railway,
            JMneNameType::Religion => JMSenseMisc::Religion,
            JMneNameType::Service => JMSenseMisc::Service,
            JMneNameType::Ship => JMSenseMisc::Ship,
            JMneNameType::Surname => JMSenseMisc::Surname,
            JMneNameType::Unclassified => JMSenseMisc::Unclassified,
        }
    }
}
