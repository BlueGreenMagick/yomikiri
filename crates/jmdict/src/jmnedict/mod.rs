mod parse;
mod types;

pub use parse::parse_jmnedict_xml;
pub use types::{JMneDict, JMneEntry, JMneKanji, JMneNameType, JMneReading, JMneTranslation};
