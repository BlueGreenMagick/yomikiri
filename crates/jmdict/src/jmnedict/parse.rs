use core::str;
use std::io::BufRead;

use log::warn;
use quick_xml::events::Event;
use quick_xml::Reader;

use crate::jmnedict::types::JMneNameType;
use crate::utils::parse_entity_enum_into;
use crate::xml::{get_next_child_in, parse_string_in_tag_into, TagName, DATE_REG};
use crate::{Error, Result};

use super::types::{JMneDict, JMneEntry, JMneKanji, JMneReading, JMneTranslation};

pub const JMNEDICT_META_ENTRY_ID: u32 = 9999990;

pub struct JMneDictParser<R: BufRead> {
    reader: Reader<R>,
    buf: Vec<u8>,
    creation_date: Option<String>,
}

impl<R: BufRead> JMneDictParser<R> {
    pub fn new(reader: R) -> Result<Self> {
        let mut reader = Reader::from_reader(reader);
        reader.config_mut().trim_text_start = true;
        reader.config_mut().trim_text_end = true;
        reader.config_mut().expand_empty_elements = true;

        let buf = Vec::new();
        let mut parser = JMneDictParser {
            reader,
            buf,
            creation_date: None,
        };
        parser.parse_jmnedict_start()?;
        Ok(parser)
    }

    pub fn next_entry(&mut self) -> Result<Option<JMneEntry>> {
        self.parse_entry_in_jmnedict()
    }

    pub fn creation_date(&self) -> Option<&str> {
        self.creation_date.as_deref()
    }

    fn parse_jmnedict_start(&mut self) -> Result<()> {
        loop {
            self.buf.clear();
            match self.reader.read_event_into(&mut self.buf)? {
                Event::Start(tag) => match tag.name().0 {
                    b"JMnedict" => {
                        return Ok(());
                    }
                    _ => {
                        warn!("Unknown global tag: {}", tag.tag_name());
                    }
                },
                Event::Eof => return Err(Error::InvalidXml("<JMnedict> not found".into())),
                _ => {}
            }
        }
    }

    fn parse_entry_in_jmnedict(&mut self) -> Result<Option<JMneEntry>> {
        loop {
            let start = match get_next_child_in(&mut self.reader, &mut self.buf, "JMnedict")? {
                Some(tag) => tag,
                None => return Ok(None),
            };
            match start.name().0 {
                b"entry" => return self.parse_in_entry().map(Some),
                _ => {
                    warn!("Unknown tag in <jmdict>: <{}>", start.tag_name());
                }
            }
        }
    }

    fn parse_in_entry(&mut self) -> Result<JMneEntry> {
        let mut entry = JMneEntry::default();
        let mut id: Option<u32> = None;

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "entry")? {
            match tag.name().0 {
                b"ent_seq" => {
                    if let Some(id) = id {
                        return Err(Error::MultipleEntryIds(id));
                    }
                    let idstr =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ent_seq")?;
                    let seq_id = str::parse::<u32>(&idstr)
                        .map_err(|_| format!("Couldn't parse as u32 number: {}", idstr))?;
                    id = Some(seq_id);
                }
                b"k_ele" => {
                    entry.kanjis.push(self.parse_in_kanji()?);
                }
                b"r_ele" => {
                    entry.readings.push(self.parse_in_reading()?);
                }
                b"trans" => {
                    entry.translations.push(self.parse_in_translation()?);
                }
                _ => {
                    warn!("Unknown tag in <entry>: <{}>", &tag.tag_name());
                }
            };
        }

        self.buf.clear();

        if let Some(id) = id {
            entry.id = id;
            if id == JMNEDICT_META_ENTRY_ID {
                self.parse_creation_date(&entry)?;
            }
            Ok(entry)
        } else {
            Err(Error::NoEntryId(self.reader.buffer_position()))
        }
    }

    fn parse_in_kanji(&mut self) -> Result<JMneKanji> {
        let mut kanji: Option<String> = None;
        let mut priorities = vec![];

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "k_ele")? {
            match tag.name().0 {
                b"keb" => {
                    if let Some(kanji) = kanji.as_ref() {
                        warn!(
                            "Warning: Found multiple <keb> in form '{}' in JMneDict",
                            kanji
                        )
                    }
                    kanji = Some(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"keb",
                    )?);
                }
                b"ke_pri" => {
                    let priority =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"ke_pri")?;
                    priorities.push(priority);
                }
                _ => {
                    warn!("Unknown tag in <k_ele>: {}", tag.tag_name());
                }
            };
        }

        let kanji = kanji.ok_or(Error::InvalidXml("No <keb> found in <k_ele>".into()))?;
        let kanji = JMneKanji {
            kanji,
            priority: priorities,
        };

        Ok(kanji)
    }

    fn parse_in_reading(&mut self) -> Result<JMneReading> {
        let mut reading: Option<String> = None;
        let mut to_kanjis = vec![];
        let mut priorities = vec![];

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "r_ele")? {
            match tag.name().0 {
                b"reb" => {
                    reading = Some(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"reb",
                    )?);
                }
                b"re_restr" => {
                    let to_form =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"re_restr")?;
                    to_kanjis.push(to_form);
                }
                b"re_pri" => {
                    let priority =
                        parse_string_in_tag_into(&mut self.reader, &mut self.buf, b"re_pri")?;
                    priorities.push(priority);
                }
                _ => {
                    warn!("Unknown tag in <r_ele>: {}", &tag.tag_name());
                }
            };
        }

        let reading = reading.ok_or(Error::InvalidXml("No <reb> found in <r_ele>".into()))?;
        let reading = JMneReading {
            reading,
            to_kanji: to_kanjis,
            priority: priorities,
        };

        Ok(reading)
    }

    fn parse_in_translation(&mut self) -> Result<JMneTranslation> {
        let mut translation = JMneTranslation::default();

        while let Some(tag) = get_next_child_in(&mut self.reader, &mut self.buf, "trans")? {
            match tag.name().0 {
                b"name_type" => {
                    parse_entity_enum_into!(
                        JMneNameType,
                        &mut self.reader,
                        &mut self.buf,
                        "name_type",
                        translation.name_type
                    );
                }
                b"xref" => {
                    translation.xref.push(parse_string_in_tag_into(
                        &mut self.reader,
                        &mut self.buf,
                        b"xref",
                    )?);
                }
                b"trans_det" => {
                    let mut is_english = true;
                    for attr in tag.attributes() {
                        let attr = attr?;
                        if attr.key.0 == b"xml:lang" {
                            let lang = attr.value;
                            if lang.as_ref() != b"eng" {
                                println!(
                                    "<trans_det> has non-english value: {}",
                                    str::from_utf8(&lang)?
                                );
                                is_english = false;
                            }
                            break;
                        }
                    }
                    if is_english {
                        let tr_text = parse_string_in_tag_into(
                            &mut self.reader,
                            &mut self.buf,
                            b"trans_det",
                        )?;
                        translation.translations.push(tr_text);
                    }
                }
                _ => {
                    warn!("Unknown tag in <trans>: {}", &tag.tag_name());
                }
            };
        }

        Ok(translation)
    }

    fn parse_creation_date(&mut self, entry: &JMneEntry) -> Result<()> {
        debug_assert_eq!(entry.id, JMNEDICT_META_ENTRY_ID);
        for trans_obj in &entry.translations {
            for translation in &trans_obj.translations {
                if let Some(date) = DATE_REG.find(translation) {
                    self.creation_date = Some(date.as_str().to_owned())
                }
            }
        }
        Ok(())
    }
}

pub fn parse_jmnedict_xml<R: BufRead>(reader: R) -> Result<JMneDict> {
    let mut parser = JMneDictParser::new(reader)?;
    let mut entries = vec![];
    while let Some(entry) = parser.next_entry()? {
        entries.push(entry);
    }
    let creation_date = parser
        .creation_date()
        .ok_or(Error::CreationDateNotFound)?
        .to_owned();
    Ok(JMneDict {
        entries,
        creation_date,
    })
}
