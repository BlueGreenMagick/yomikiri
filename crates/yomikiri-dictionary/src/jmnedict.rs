use std::collections::HashMap;

use yomikiri_jmdict::jmdict::JMSenseMisc;
use yomikiri_jmdict::jmnedict::{JMneKanji, JMneNameType, JMneReading, JMneTranslation};

use crate::entry::{GroupedSense, Rarity, WordEntryInner};
use crate::name::{GroupedNameItem, NameEntry, NameItem, NameType};
use crate::{Kanji, PartOfSpeech, Reading, Result, Sense, WordEntry};

type NameEntryKanji = String;

#[derive(Debug, Clone, PartialEq, Eq)]
struct NameEntryFragmentValue {
    id: u32,
    reading: String,
    // TODO: Instead of having a vector everywhere, reuse combinations.
    name_type: Vec<NameType>,
    priority: u16,
}

/// Parses JMnedict xml and returns (Vec<NameEntry>, Vec<Entry[]>)
/// where the second `Entry[]` is treated as a 'word entry'.
///
/// A `JMneTranslation` entry either creates (multiple) `NameItem`, or one `Sense` in `Entry`, not both.
///
/// A `NameEntry` is created for JMnedict entry with at least 1 kanji, and is a 'name-type' (of a person)
/// where its translation is only a transliteration of the reading.
///
/// Whether a `JMneTranslation` is a 'name-type' is determined by its `name_type` property.
/// 4 variants `Forename`, `Surname`, `Female`, `Male` are considered name-type, and creates `NameItem`.
/// Other variants create a `Sense` in a `Entry`.
///
/// If `name_type` of `JMneTranslation` has types of both kind, then
///     - If it is `Person` with `Female` or `Male`, only a WordItem is created, and `Female` / `Male` is ignored.
///       The gender tag seems to be mistagged here to only indicate the person's gender, instead of the forename + gender info.
///     - For other tag combinations, we only create `NameItem`, as it's a transliteration anyway.
pub fn parse_jmnedict_xml(xml: &str) -> Result<(Vec<NameEntry>, Vec<WordEntry>)> {
    let mut fragments: HashMap<NameEntryKanji, Vec<NameEntryFragmentValue>> = HashMap::new();
    let mut name_entries: Vec<NameEntry> = vec![];
    let mut word_entries: Vec<WordEntry> = vec![];
    let jmnedict = yomikiri_jmdict::jmnedict::parse_jmnedict_xml(xml)?;

    for entry in jmnedict.entries {
        let mut for_names: Vec<JMneTranslation> = vec![];
        let mut for_words: Vec<JMneTranslation> = vec![];
        if entry.kanjis.is_empty() {
            for_words.extend(entry.translations);
        } else {
            for translation_obj in entry.translations {
                if is_name_type(&translation_obj.name_type) {
                    for_names.push(translation_obj);
                } else {
                    for_words.push(translation_obj);
                }
            }
        }

        if !for_names.is_empty() {
            for kanji in &entry.kanjis {
                for reading in &entry.readings {
                    if !reading.is_for_kanji(&kanji.kanji) {
                        continue;
                    }
                    for trans_obj in &for_names {
                        let fragment_value = NameEntryFragmentValue {
                            id: entry.id,
                            reading: reading.reading.clone(),
                            name_type: trans_obj.name_type.clone(),
                            priority: 0, // TODO: Set priority info
                        };
                        if !fragments.contains_key(&kanji.kanji) {
                            fragments.insert(kanji.kanji.clone(), vec![]);
                        }
                        fragments
                            .get_mut(&kanji.kanji)
                            .unwrap()
                            .push(fragment_value)
                    }
                }
            }
        }

        if !for_words.is_empty() {
            let inner =
                WordEntryInner::from_jmnedict(entry.id, &entry.kanjis, &entry.readings, &for_words);
            let entry = WordEntry::new(inner)?;
            word_entries.push(entry);
        }
    }

    for (kanji, values) in fragments {
        let mut groups: Vec<GroupedNameItem> = vec![];
        for val in values {
            for grp in &mut groups {
                if grp.types == val.name_type {
                    grp.items.push(NameItem {
                        id: val.id,
                        reading: val.reading,
                    });
                    break;
                }
            }
        }
        let name_entry = NameEntry { kanji, groups };
        name_entries.push(name_entry);
    }

    Ok((name_entries, word_entries))
}

/// Returns `true` if it is a 'name-type' suitable for a NameEntry.
fn is_name_type(name_types: &[JMneNameType]) -> bool {
    let mut has_fore_surname = false;
    let mut has_gender = false;
    let mut has_person = false;
    for kind in name_types {
        match kind {
            JMneNameType::Forename | JMneNameType::Surname => {
                has_fore_surname = true;
            }
            JMneNameType::Female | JMneNameType::Male => {
                has_gender = true;
            }
            JMneNameType::Person => {
                has_person = true;
            }
            _ => {}
        };
    }

    if has_gender && has_person {
        false
    } else {
        has_fore_surname || has_gender
    }
}

impl WordEntryInner {
    fn from_jmnedict(
        id: u32,
        kanjis: &[JMneKanji],
        readings: &[JMneReading],
        translations: &[JMneTranslation],
    ) -> WordEntryInner {
        let kanjis = kanjis
            .iter()
            .map(|k| Kanji {
                kanji: k.kanji.clone(),
                rarity: Rarity::Normal,
            })
            .collect();
        let readings = readings
            .iter()
            .map(|r| Reading {
                reading: r.reading.clone(),
                to_kanji: r.to_kanji.clone(),
                nokanji: false,
                rarity: Rarity::Normal,
            })
            .collect();
        let grouped_senses = vec![GroupedSense {
            pos: vec![PartOfSpeech::Noun],
            senses: translations
                .iter()
                .map(|t| Sense {
                    meanings: t.translations.clone(),
                    to_kanji: vec![],
                    to_reading: vec![],
                    misc: t.name_type.iter().map(|t| JMSenseMisc::from(*t)).collect(),
                    info: vec![],
                    dialects: vec![],
                })
                .collect(),
        }];

        WordEntryInner {
            id,
            kanjis,
            readings,
            grouped_senses,
            priority: 0, // TODO: Calculate priority from entry
        }
    }
}
