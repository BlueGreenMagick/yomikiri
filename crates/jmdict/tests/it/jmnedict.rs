use insta::assert_yaml_snapshot;
use yomikiri_jmdict::jmnedict::parse_jmnedict_xml;
use yomikiri_jmdict::Result;

#[test]
fn test_parse_jmnedict() -> Result<()> {
    let xml = r#"<?xml version="1.0"?>
<!DOCTYPE JMnedict [
    <!ELEMENT JMnedict (entry*)>
<!ENTITY char "character">
<!ENTITY company "company name">
]>
<!-- JMnedict created: 2024-08-29 -->
<JMnedict>
<entry>
<ent_seq>5000000</ent_seq>
<k_ele>
<keb>ゝ泉</keb>
</k_ele>
<r_ele>
<reb>ちゅせん</reb>
</r_ele>
<trans>
<name_type>&given;</name_type>
<trans_det>Chusen</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5000001</ent_seq>
<k_ele>
<keb>〆</keb>
</k_ele>
<r_ele>
<reb>しめ</reb>
</r_ele>
<trans>
<name_type>&fem;</name_type>
<trans_det>Shime</trans_det>
</trans>
</entry>
<entry>
<ent_seq>5744957</ent_seq>
<k_ele>
<keb>全日本労働総同盟</keb>
</k_ele>
<r_ele>
<reb>ぜんにほんろうどうそうどうめい</reb>
<re_pri>spec1</re_pri>
</r_ele>
<trans>
<name_type>&organization;</name_type>
<trans_det>Japanese Confederation of Labor (1936-1939)</trans_det>
<trans_det>Zensō</trans_det>
</trans>
<trans>
<name_type>&organization;</name_type>
<trans_det>Japanese Confederation of Labor (1964-1987)</trans_det>
<trans_det>Dōmei</trans_det>
</trans>
</entry>
</JMnedict>"#;
    let result = parse_jmnedict_xml(&xml)?;
    assert_yaml_snapshot!(result);
    Ok(())
}
