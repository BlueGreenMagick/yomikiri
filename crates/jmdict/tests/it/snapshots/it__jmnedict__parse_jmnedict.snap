---
source: crates/jmdict/tests/it/jmnedict.rs
assertion_line: 76
expression: result
---
entries:
  - id: 5000000
    kanjis:
      - kanji: ゝ泉
        priority: []
    readings:
      - reading: ちゅせん
        to_kanji: []
        priority: []
    translations:
      - name_type:
          - forename
        xref: []
        translations:
          - Chusen
  - id: 5000001
    kanjis:
      - kanji: 〆
        priority: []
    readings:
      - reading: しめ
        to_kanji: []
        priority: []
    translations:
      - name_type:
          - forename (female)
        xref: []
        translations:
          - Shime
  - id: 5744957
    kanjis:
      - kanji: 全日本労働総同盟
        priority: []
    readings:
      - reading: ぜんにほんろうどうそうどうめい
        to_kanji: []
        priority:
          - spec1
    translations:
      - name_type:
          - organization
        xref: []
        translations:
          - Japanese Confederation of Labor (1936-1939)
          - Zensō
      - name_type:
          - organization
        xref: []
        translations:
          - Japanese Confederation of Labor (1964-1987)
          - Dōmei
  - id: 9999990
    kanjis:
      - kanji: ＪＭｎｅｄｉｃｔ
        priority: []
    readings:
      - reading: ジェイエムエヌイーディクト
        to_kanji: []
        priority: []
    translations:
      - name_type:
          - artwork
        xref: []
        translations:
          - "Japanese-Multilingual Named Entity Dictionary Project - Creation Date: 2024-08-29"
creation_date: 2024-08-29
